thoth-app/src/component/work.rs

```diff
@@ -389,11 +411,20 @@ impl Component for WorkComponent {
                 self.work.edition.neq_assign(edition)
             }
             Msg::ChangeDoi(value) => {
-                let doi = match value.trim().is_empty() {
-                    true => None,
-                    false => Some(value.trim().to_owned()),
-                };
-                self.work.doi.neq_assign(doi)
+                if self.doi.neq_assign(value.trim().to_owned()) {
+                    // If DOI is not correctly formatted, display a warning.
+                    // If no DOI was provided, no format check is required.
+                    // Don't update self.work.doi yet, as user may later
+                    // overwrite a new valid value with an invalid one.
+                    if !self.doi.is_empty() && self.doi.parse::<Doi>().is_err() {
+                        self.doi_warning = self.doi.parse::<Doi>().unwrap_err().to_string();
+                    } else {
+                        self.doi_warning.clear();
```

Comment:
I guess we want to set the value here, once it's been parsed, so that the field displays the non-URI version of the DOI. Otherwise I'm finding that inputting a valid DOI (in URI format) results in the following:

![Screenshot_2021-06-10_14-54-40](https://user-images.githubusercontent.com/20479024/121537412-cee4af80-c9fb-11eb-8c2e-7295b2fc73be.png)

Reply:
It was tricky to work out a behaviour for this field that wouldn't be either confusing or error-prone. I decided the best compromise was only setting the `self.work.doi` value on save, so if you pressed Save at this point, then the field would change to display the non-URI version. The problem with setting `self.work.doi` as soon as a parsable value is found is that the user might then continue typing and change the valid DOI to an invalid one - if so, then pressing Save would overwrite the original value with the intermediate valid value. It could also be confusing to the user if the value they typed in suddenly changed while they were still typing.

There are probably alternative approaches that would get round those issues, but my feeling is that if someone's going to ignore the static button and input a full URI, they probably won't notice (or care) that the formatting remains odd-looking until they save!

Reply:
Oh, I didn't mean setting `self.work.doi`, but setting the parsed value of `self.doi`, e.g. `self.doi = self.doi.parse::<Doi>().unwrap().to_string();`. That way if you paste the full URI it'll get formatted properly. Something like:

```rust
                    self.doi_warning.clear();
                    match self.doi.parse::<Doi>() {
                        Err(e) => self.doi_warning = e.to_string(),
                        Ok(value) => self.doi = value.to_string(),
                    }
                    true
```

Reply:
Well, the above, but with your `is_empty()` checks...

Reply:
It might be worth expanding the list of errors, the empty check could be done at parsing, outputting an `IdentifierParseError` when it fails or an e.g. `EmptyIdentifier` if the input is empty. That way you can match the error, e.g.

```rust
                    self.doi_warning.clear();
                    match self.doi.parse::<Doi>() {
                        Err(e) => {
                            match e {
                                ThothError::IdentifierParseError(_, _) => self.doi_warning = e.to_string(),
                                _ => {}
                            }
                        },
                        Ok(value) => self.doi = value.to_string(),
                    }
                    true
```
With `thoth-api/src/model/mod.rs:105`:

```rust
            if input.is_empty() {
                Err(ThothError::EmptyIdentifier("DOI".to_string()))
            } else {
                Err(ThothError::IdentifierParseError(
                    input.to_string(),
                    "DOI".to_string(),
                ))
            }
```